model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java

```diff
@@ -0,0 +1,273 @@
+    public void setName(String name) {
+        checkEntityVersionForUpdate();
+        metadata.setName(name);
+    }
+
+    @Override
+    public void setDescription(String description) {
+        checkEntityVersionForUpdate();
+        metadata.setDescription(description);
+    }
+
+    @Override
+    public Set<String> getCompositeRoles() {
+        return metadata.getCompositeRoles();
+    }
+
+    @Override
+    public void setCompositeRoles(Set<String> compositeRoles) {
+        checkEntityVersionForUpdate();
+        metadata.setCompositeRoles(compositeRoles);
+    }
+
+    @Override
+    public void addCompositeRole(String roleId) {
+        checkEntityVersionForUpdate();
+        metadata.addCompositeRole(roleId);
+    }
+
+    @Override
+    public void removeCompositeRole(String roleId) {
+        checkEntityVersionForUpdate();
+        metadata.removeCompositeRole(roleId);
+    }
+
+    @Override
+    public Map<String, List<String>> getAttributes() {
+        Map<String, List<String>> result = new HashMap<>();
+        for (JpaRoleAttributeEntity attribute : attributes) {
+            List<String> values = result.getOrDefault(attribute.getName(), new LinkedList<>());
+            values.add(attribute.getValue());
+            result.put(attribute.getName(), values);
+        }
+        return result;
+    }
+
+    @Override
+    public List<String> getAttribute(String name) {
+        return attributes.stream()
+                .filter(a -> Objects.equals(a.getName(), name))
+                .map(JpaRoleAttributeEntity::getValue)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public void setAttributes(Map<String, List<String>> attributes) {
+        checkEntityVersionForUpdate();
+        for (Iterator<JpaRoleAttributeEntity> iterator = this.attributes.iterator(); iterator.hasNext();) {
+            JpaRoleAttributeEntity attr = iterator.next();
+            iterator.remove();
+            attr.setRole(null);
+        }
+        if (attributes != null) {
+            for (Map.Entry<String, List<String>> entry : attributes.entrySet()) {
+                setAttribute(entry.getKey(), entry.getValue());
+            }
+        }
+    }
+
+    @Override
+    public void setAttribute(String name, List<String> values) {
+        checkEntityVersionForUpdate();
+        removeAttribute(name);
+        for (String value : values) {
+            JpaRoleAttributeEntity attribute = new JpaRoleAttributeEntity(this, name, value);
+            attributes.add(attribute);
+        }
+    }
+
+    @Override
+    public void removeAttribute(String name) {
+        checkEntityVersionForUpdate();
+        for (Iterator<JpaRoleAttributeEntity> iterator = attributes.iterator(); iterator.hasNext();) {
+            JpaRoleAttributeEntity attr = iterator.next();
+            if (Objects.equals(attr.getName(), name)) {
+                iterator.remove();
+                attr.setRole(null);
+            }
+        }
+    }
```

Comment:
While the setters store their data in metadata as well, I don't see this happening for the attribute field (all setter/remove instances).
For example for `removeAttribute` I would have expected a call to `metadata.removeAttribute(name);` 
Is this missing here, or is this triggered somewhere else?

Reply:
Attributes are stored as [association](https://github.com/vramik/keycloak/blob/KEYCLOAK-18565-roles/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/entity/JpaRoleEntity.java#L67-L68), therefore the field is omitted from json when [serialization](https://github.com/keycloak/keycloak/blob/main/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/hibernate/jsonb/JsonbType.java#L78) is done (the same applies to id field).

Reply:
Thanks for the explanation, I can now understand. Still it was a "surprise" for me, maybe that's something that would help future developers if it would be part of a general rule for this store (unless a field is being migrated to a searchable field in a future version, and will be in metadata so that the previous entity version finds it there as well when reading the data).

Consider this comment to be resolved.